keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1085 @@
+
+#### Asynchronous Errors/Failures
+
+This refers to a class of errors that could happen after the StatefulSet controller has successfully performed a `patch` operation on the PVC object. These errors are asynchronous in nature and can happen at any time after the `patch` operation. In the diagram above, it refers to the arrow denoted with `(10)`. (Though in reality, `kubelet` is involved in the resize as well, when an online expansion is performed, but we leave it out of the diagram).
+
+* The CSI driver doesn't support online expansions.
+  * This is a scenario where the driver cannot do an expansion while the volume is published to a node/the pod that's using it is up. When this happens, the CSI driver will return an error when the CSI resizer calls it (eg `FailedPrecondition`). In this proposal we won't be supporting the scenario of not supporting online expansions. Ideally, we'd want to know ahead of time (during StatefulSet admission/validation) if the CSI driver supports online expansion or not - but it might not be so simple. We'll discuss it in details when we talk about [Validation](#validation).
+* The CSI driver doesn't support offline expansions.
+  * This is a scenario where the volume must be published to a node/the pod that's using it must be up in order to perform an expansion. This is not exactly an error. The expansion process will be stuck until the volume is published on a node. This could happen in scenarios where a user attempts to resize a PVC that's not currently attached to a pod (eg a PVC is bound to a pod of the StatefulSet and then the user downscales the StatefulSet). We'll have to make sure to not get in the way of the user when try try to recover from this scenario.
+* No space left/exceeding quota
+  * This is a scenario where the driver returns an error to the resizer (eg `ResourceExhaused`/`OutOfRange`) because there's no space left for the resize process, or the user exceeds a quota on the provider they are using for storage. There is nothing we can do in this scenario other than not getting in the way of the user when they try to recover from it, and, continue reconciliation once they apply successful recovery. The recovery could involve steps like
+    * Using the mechanisms that are described in [KEP 1790](../1790-recover-resize-failure/README.md)
+    * Requesting more quota from the provider
+  * It's important to note, that it's possible for partial failures to occur as well - which might be more tricky for the user to handle. For example, a user could do a resize in a StatefulSet that has 3 replicas and the resize could succeed on 2 of them and fail only on the last.
+    * In such case, the user won't be able to apply the mechanisms in [KEP 1790](../1790-recover-resize-failure/README.md) and their only recourse would be to request more quota from the provider, or downsize their StatefulSet (number of replicas).
```

Comment:
If this happens - are we going to block rollout of other changes to statefulset? Say user increased volume size from 10Gi to 20Gi and we could only update 2 replicas, what would be the state of statefulset in this case. Can user still keep using it? For volume expansion failures - generally expansion failures do not prevent a volume from being mounted or attached.

Reply:
Since this is an asynchronous error, it won't block the statefulset reconciliation logic. The user will still be able to modify the pod spec (and have it take effect) and even resize volume again (which may succeed for the first 2 replicas as long as the resize doesn't make their PVC exceed quota). 
The only effect that the failed resize on the third replica would have, besides an error event on the PVC object iself, is the `finishedReconciliationGeneration` (part of the feedback mechanism) lag behind the generation of the statefulset and `readyReplicas` be set to 2 instead of 3, until the user resolves the resize issue
I can make that more clear in the doc

Reply:
I am confused. The KEP says elsewhere:

> Another reason is that we want for the PVC reconciliation to match the cadence of pod creation/modification in the StatefulSet. If the user uses monotonic (`OrderedReadyPodManagement`) roll out strategy, we'd want the PVC reconciliation process to respect that strategy.

So what really happens with an  `OrderedReadyPodManagement` StatefulSet, whose Pod template + PVC template is updated in a single PATCH? Will Pods be restarted + PVC resized in lock-step (Pod1 gets Ready - PVC1 is resized - Pod2 gets ready - PVC2 is resizer- ...) or can Pod2 be upgraded when PVC1 gets stuck at resizing?

Reply:
@jsafrane the patching of the PVC will happen in lock step with the pods getting healthy (not necessarily ready). so once pod1 gets ready, we'll patch PVC1. once pod2 gets ready, we'll patch PVC2, etc...
we are not going to wait on PVCs being fully reconciled (capacity matches request storage size) though. Pod 2 will be upgraded if PVC1 encounters an error during resize (an error that wasn't caught during patching, described in https://github.com/kubernetes/enhancements/blob/95a7fe96a22b2c3ba467df2be9879439d23c5d3f/keps/sig-storage/661-statefulset-volume-resize/README.md#asynchronous-errorsfailures)